cdap-master/src/main/java/co/cask/cdap/master/startup/MasterStartupTool.java

```diff
@@ -147,10 +155,29 @@ private static Injector createInjector(CConfiguration cConf, Configuration hConf
     return Guice.createInjector(
       new ConfigModule(cConf, hConf),
       new ZKClientModule(),
-      new LocationRuntimeModule().getDistributedModules(),
       new IOModule(),
       new KafkaClientModule(),
-      new DiscoveryRuntimeModule().getDistributedModules()
+      new DiscoveryRuntimeModule().getDistributedModules(),
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          bind(LocationFactory.class).to(RootLocationFactory.class);
+        }
+
+        @Provides
+        @Singleton
+        private RootLocationFactory providesLocationFactory(final Configuration hConf, CConfiguration cConf) {
```

Comment:
Turn this method into a Guice `Provider` so that you can use it in both `LocationRuntimeModule` as well as in here.

```java
bind(LocationFactory.class).toProvider(RootLocationFactoryProvider.class).in(Scopes.SINGLETON);
....

public class RootLocationFactoryProvider implements Provider<RootLocationFactory> {
  @Inject
  RootLocationFactoryProvider(CConfiguration cConf, Configuration hConf) {
     ...
  }

  @Override
  public RootLocationFactory get() {
     ...
  }
}
```